fuel-vm/src/interpreter/blockchain.rs

```diff
@@ -257,161 +569,166 @@ where
-            ref memory,
-            ref mut registers,
-            ..
-        } = self;
+pub(crate) struct StateWordCtx<'vm, S> {
+    pub storage: &'vm mut S,
+    pub memory: &'vm [u8; MEM_SIZE],
+    pub context: &'vm Context,
+    pub fp: Reg<'vm, FP>,
+    pub pc: RegMut<'vm, PC>,
+}
 
-        state_clear_qword(&contract_id, storage, memory, &mut registers[rb], input)?;
+pub(crate) fn state_read_word<S: InterpreterStorage>(
+    StateWordCtx {
+        storage,
+        memory,
+        context,
+        fp,
+        pc,
+    }: StateWordCtx<S>,
+    result: &mut Word,
+    got_result: &mut Word,
+    c: Word,
+) -> Result<(), RuntimeError> {
+    let cx = checked_add_word(c, Bytes32::LEN as Word)?;
 
-        self.inc_pc()
+    if cx > VM_MAX_RAM {
+        return Err(PanicReason::MemoryOverflow.into());
     }
 
-    pub(crate) fn state_read_word(&mut self, ra: RegisterId, rb: RegisterId, c: Word) -> Result<(), RuntimeError> {
-        Self::is_register_writable(ra)?;
-        Self::is_register_writable(rb)?;
-
-        let cx = checked_add_word(c, Bytes32::LEN as Word)?;
-
-        if cx > VM_MAX_RAM {
-            return Err(PanicReason::MemoryOverflow.into());
-        }
-
-        let (c, cx) = (c as usize, cx as usize);
+    let (c, cx) = (c as usize, cx as usize);
 
-        let contract = self.internal_contract()?;
+    let contract = internal_contract(context, fp, memory)?;
 
-        // Safety: Memory bounds are checked by the interpreter
-        let key = unsafe { Bytes32::as_ref_unchecked(&self.memory[c..cx]) };
+    // Safety: Memory bounds are checked by the interpreter
+    let key = unsafe { Bytes32::as_ref_unchecked(&memory[c..cx]) };
 
-        let result = self
-            .storage
-            .merkle_contract_state(contract, key)
-            .map_err(RuntimeError::from_io)?
-            .map(|state| unsafe { Bytes8::from_slice_unchecked(state.as_ref().as_ref()).into() })
-            .map(Word::from_be_bytes);
+    let value = storage
+        .merkle_contract_state(contract, key)
+        .map_err(RuntimeError::from_io)?
+        .map(|state| unsafe { Bytes8::from_slice_unchecked(state.as_ref().as_ref()).into() })
+        .map(Word::from_be_bytes);
 
-        self.registers[ra] = result.unwrap_or(0);
-        self.registers[rb] = result.is_some() as Word;
+    *result = value.unwrap_or(0);
+    *got_result = value.is_some() as Word;
 
-        self.inc_pc()
-    }
+    inc_pc(pc)
+}
 
-    pub(crate) fn state_read_qword(&mut self, a: Word, rb: RegisterId, c: Word, d: Word) -> Result<(), RuntimeError> {
-        Self::is_register_writable(rb)?;
-        let contract_id = *self.internal_contract()?;
-        let input = StateReadQWord::new(a, c, d, self.ownership_registers())?;
-        let Self {
-            ref storage,
-            ref mut memory,
-            ref mut registers,
-            ..
-        } = self;
+pub(crate) fn state_write_word<S: InterpreterStorage>(
+    StateWordCtx {
```

Comment:
didn't realize it was possible to destructure inside fn args, this is nice way to reduce boilerplate